File: rust/src/leetcode/Q200_number_of_islands.rs
/*
 * @lc app=leetcode id=200 lang=rust
 *
 * [200] Number of Islands
 */

// @lc code=start
impl Solution {
    fn dfs(row: i32, col: i32, grid: &mut Vec<Vec<char>>) -> bool {
        if row < 0
            || col < 0
            || row >= grid.len() as i32
            || col >= grid[0].len() as i32
            || grid[row as usize][col as usize] == '0'
        {
            return false;
        }

        grid[row as usize][col as usize] = '0';

        Self::dfs(row - 1, col, grid);
        Self::dfs(row + 1, col, grid);
        Self::dfs(row, col - 1, grid);
        Self::dfs(row, col + 1, grid);

        true
    }

    pub fn num_islands(grid: Vec<Vec<char>>) -> i32 {
        let mut result: i32 = 0;
        let mut grid: Vec<Vec<char>> = grid;

        for i in 0..grid.len() {
            for j in 0..grid[0].len() {
                if Self::dfs(i as i32, j as i32, &mut grid) {
                    result += 1;
                }
            }
        }

        result
    }
}
// @lc code=end

File: python/Q200_number_of_islands.py
#
# @lc app=leetcode id=200 lang=python3
#
# [200] Number of Islands
#


# @lc code=start
class Solution:
    def dfs(self, row: int, col: int, new_island: bool = False) -> None:
        if (
            row < 0
            or col < 0
            or row >= self.ROWS
            or col >= self.COLS
            or self.grid[row][col] == "0"
        ):
            return

        if new_island:
            self.islands += 1

        self.grid[row][col] = "0"

        for x, y in self.directions:
            self.dfs(row + x, col + y)

    def numIslands(self, grid: list[list[str]]) -> int:
        self.ROWS: int = len(grid)
        self.COLS: int = len(grid[0])
        self.grid: list[list[str]] = grid

        self.islands: int = 0

        self.directions: list[tuple[int, int]] = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        for x in range(self.ROWS):
            for y in range(self.COLS):
                self.dfs(x, y, True)

        return self.islands

# @lc code=end
